### ISSUE TITLE
Inserts extra closing quote in some cases

### ISSUE THREAD

**Issue by viridia** [atom/bracket-matcher#298]
### Description

Typing a single closing quote inserts two quote marks, depending on the text that follows between the opening quote and the closing quote.

### Steps to Reproduce

Given a line of text such as:

    this is a '{test} line.

where the cursor is located immediately after the '}'. If you type a single quote character, you'll get this:

    this is a '{test}'' line.

However, if you have a line like this:

    this is a 'test line.

And you type a closing quote, you'll get a single closing quote as you would expect:

    this is a 'test' line.

**Expected behavior:**

I would expect if there close quote is preceded by a matching opening quote (i.e. there's an odd number of quote characters preceding the current position) that only a single character would be inserted.

**Actual behavior:** [What actually happens]

It inserts two quote characters, as if there were no preceding matching quote.

**Reproduces how often:**

This is 100% reproduceable for me.

### Versions

Atom    : 1.18.0
Electron: 1.3.15
Chrome  : 52.0.2743.82
Node    : 6.5.0
apm  1.18.1
npm  3.10.10
node 6.9.5 x64
python 2.7.10
git 2.10.2

### Additional Information

This has been driving me nuts ever since I started using Atom two years ago, I finally tracked down what the exact steps were. It's extra-annoying because you can't use backspace to delete one of the quotes, it auto-deletes both of them. To insert a single matching quote you have to cursor past the second inserted closing quote and then hit backspace.

**Comment by 50Wliu** [comment]
This is because there's a word-boundary check when deciding whether or not to close the existing bracket.
https://github.com/atom/bracket-matcher/blob/9db77b302965aa38de8d29335765d33e6bf70ef0/lib/bracket-matcher.coffee#L50

**Comment by viridia** [comment]
Well, whatever the reason, the behavior is terribly annoying when writing code, especially when writing ES6 template strings in JS. Instead of the bracket-matcher saving you keystrokes, it means that you end up having to type more keystrokes than you would otherwise.

**Comment by 50Wliu** [comment]
Yeah, I agree.  Just haven't had time to work out a better solution.

**Comment by viridia** [comment]
I don't know of a perfect answer (not sure a perfect answer is even possible without actually parsing the text being edited), but the thing I'd try is to simply count the number of preceding quote marks on the line;

* If it's an even number (meaning that each preceding quoted string has a matching open and close quote), then do the current behavior which looks at the characters adjacent to the cursor;
* If it's odd (meaning that there's an unmatched open quote) then suppress the auto-insertion and simply insert a single quote character.

This assumes:

* That this would only be true for ASCII single and double quote, not other kinds of brackets or smart quotes;
* That most languages don't allow line breaks in single or double quotes (although plain text of course does allow it, but that shouldn't be a problem since it would be rare to have one quoted section end and another begin on the same line).
